lib/web/service-worker/service-worker-container.js

```diff
@@ -0,0 +1,201 @@
+import { Worker, MessageChannel } from 'node:worker_threads'
+import { DeferredPromise } from '@open-draft/deferred-promise'
+import { ServiceWorkerRegistration } from './service-worker-registration.js'
+import { ServiceWorker } from './service-worker.js'
+import { parseModuleUrlFromStackTrace } from './utils/parseModuleUrlFromStackTrace.js'
+import { interceptor } from './interceptor.js'
+
+/**
+ * @typedef {Object} WorkerData
+ * @property {string} scriptUrl
+ * @property {unknown} options @todo
+ * @property {import('./client.js').SerializedClient} clientInfo
+ * @property {MessagePort} clientMessagePort
+ * @property {MessagePort} interceptorMessagePort
+ */
+
+const clientMessageChannel = new MessageChannel()
+const interceptorMessageChannel = new MessageChannel()
+
+/**
+ * @see https://w3c.github.io/ServiceWorker/#serviceworkercontainer-interface
+ */
+export class ServiceWorkerContainer {
+  /** @type {ServiceWorkerRegistration | undefined} */
+  #registration
+  /** @type {DeferredPromise<ServiceWorkerRegistration>} */
+  #ready
+
+  /** @todo Event handlers (oncontrollerchange, onmessage, onerror) */
+
+  constructor () {
+    this.#ready = new DeferredPromise()
+  }
+
+  /**
+   * @returns {Promise<ServiceWorkerRegistration>}
+   *
+   * @see https://w3c.github.io/ServiceWorker/#navigator-service-worker-ready
+   */
+  get ready () {
+    return this.#ready
+  }
+
+  /**
+   * @returns {ServiceWorker | null}
+   *
+   * @see https://w3c.github.io/ServiceWorker/#navigator-service-worker-controller
+   */
+  get controller () {
+    return this.#registration?.active || null
+  }
+
+  /**
+   * @param {string} scriptUrl
+   * @param {ServiceWorkerRegistrationOptions} [options]
+   * @returns {Promise<ServiceWorkerRegistration>}
+   *
+   * @see https://w3c.github.io/ServiceWorker/#navigator-service-worker-register
+   */
+  async register (scriptUrl, options = {}) {
+    /** @type {SerializedClient} */
+    const clientInfo = {
+      id: process.pid.toString(),
+      url: parseModuleUrlFromStackTrace(new Error()),
+      type: 'worker',
+      frameType: '???' /** @todo */
+    }
+
+    const worker = new Worker(new URL('./worker.ts', import.meta.url), {
```

Comment:
The container (the `serviceWorker` object) spawns a new worker of `worker.js` to provide scope isolation akin to how a Service Worker operates in the browser. I suggest we keep that isolation for the worker code not to leak into the main thread.